pinot-core/src/main/java/org/apache/pinot/core/query/reduce/resultsetter/DistinctResultSetter.java

```diff
@@ -0,0 +1,128 @@
+    // inside a DataTable
+    Collection<DataTable> dataTables = dataTableMap.values();
+    Preconditions.checkState(dataSchema.size() == 1, "DataTable from server for DISTINCT should have exactly one row");
+    Preconditions.checkState(dataSchema.getColumnDataType(0) == DataSchema.ColumnDataType.OBJECT,
+        "DistinctAggregationFunction should return result of type OBJECT");
+    Object mergedIntermediateResult = null;
+    // go over all the data tables from servers
+    for (DataTable dataTable : dataTables) {
+      Preconditions.checkState(dataTable.getNumberOfRows() == 1);
+      // deserialize the DistinctTable
+      Object intermediateResultToMerge = dataTable.getObject(0, 0);
+      Preconditions.checkState(intermediateResultToMerge instanceof DistinctTable);
+      DistinctTable distinctTable = (DistinctTable) intermediateResultToMerge;
+      // since DistinctTable uses the Table interface and during deserialization, we didn't
+      // have all the necessary information w.r.t ORDER BY, limit etc, we set it now
+      // before merging so that resize/trimming/sorting happens correctly
+      distinctTable.addLimitAndOrderByInfo(_brokerRequest);
+      if (mergedIntermediateResult == null) {
+        mergedIntermediateResult = intermediateResultToMerge;
+      } else {
+        _aggregationFunction.merge(mergedIntermediateResult, intermediateResultToMerge);
+      }
+    }
+
+    DistinctTable distinctTable = (DistinctTable) mergedIntermediateResult;
+    // finish the merging, sort (if ORDER BY), get iterator
+    distinctTable.finish(true);
+
+    List<Serializable[]> resultSet = new ArrayList<>(distinctTable.size());
+    String[] columnNames = distinctTable.getDataSchema().getColumnNames();
+    Iterator<Record> iterator = distinctTable.iterator();
+    while (iterator.hasNext()) {
+      Record record = iterator.next();
+      Object[] columns = record.getValues();
+      Serializable[] distinctRow = new Serializable[columns.length];
+      for (int col = 0; col < columns.length; col++) {
+        final Serializable columnValue = AggregationFunctionUtils.getSerializableValue(columns[col]);
+        distinctRow[col] = columnValue;
+      }
+      resultSet.add(distinctRow);
+    }
+
+    // Up until now, we have treated DISTINCT similar to another aggregation function even in terms
+    // of the result from function and merging results.
+    // However, the DISTINCT query is just another SELECTION style query from the user's point
+    // of view and will return one or records in the result table for the column(s) selected and so
+    // for that reason, response from broker should be a selection query result.
+    brokerResponseNative.setSelectionResults((new SelectionResults(Arrays.asList(columnNames), resultSet)));
+  }
+
+  private String[] getDistinctColumns() {
+    return _brokerRequest.getAggregationsInfo().get(0).getAggregationParams().get("column").split(":");
```

Comment:
(nit) I think it is better to use the constants for ":" and and "column". These constant are already defined in FunctionCallAstNode and are used in multiple places wherever we need to get the set of columns that are arguments to DISTINCT.

Reply:
fixed.